iree/turbine/kernel/wave/index_sequence_analysis.py

```diff
@@ -125,10 +113,135 @@ def has_strided_access(node: fx.Node) -> bool:
                     )
                     for j, dim in enumerate(custom.register_type.symbolic_shape)
                 }
+                ops_to_combine.append(write)
 
+        # Useful to handle write/read dependency
+        custom.replace_all_uses_with(ops_to_combine)
         custom.graph.erase_node(operator)
 
 
+def partition_ops_with_gpr_offsets(trace: CapturedTrace, constraints: list[Constraint]):
+    """
+    This function analyzes the index sequence of reads and writes in a graph.
+    If the reads or writes have incontiguous offsets based on GPR_NUM, we'd
+    need to split these reads/writes appropriately.
+
+    e.g a vector<16xf16> may be owned by lane 0, and lane 16 in this layout:
+    [0, 0, 0, 0, 16, 16, 16, 16, 0, 0, 0, 0, 16, 16, 16, 16].
+
+    With our current gloassary, this means we have 2 VGPR "chunks".
+    [0:4) and [8:12) for lane0, and [4:8) and [12:16) for lane16.
+    To the lane it should just look like vector<8xf16>.
+    Hence for this example, we'd need two reads of vector<4xf16> and a couple
+    insert_slices to combine them together into a single vector<8xf16>.
+
+    """
+
+    def has_gpr_offsets(node: fx.Node) -> bool:
+        """
+        Checks for writes on 2d tensors with strided access on a single dimension that
+        read more than a single element.
+        """
+        custom = get_custom(node)
+        if not isinstance(custom, (Read, Write)):
+            return False
+        dims_with_gpr_offset = [
+            v.start for k, v in custom.index.items() if v.start.has(GPR_NUM)
+        ]
+        if not dims_with_gpr_offset:
+            return False
+        num_dims_with_gpr_offsets = len(dims_with_gpr_offset)
+        if num_dims_with_gpr_offsets > 1:
+            raise NotImplementedError("Currently only handle 1 dim with gpr offset.")
+        return True
+
+    strided_operators = trace.walk(has_gpr_offsets)
+    for operator in strided_operators:
+        custom = get_custom(operator)
+        simplified_index = {
+            dim: simplify_index(custom.index.get(dim, custom.index[dim]))
+            for dim in custom.index
+        }
+        elements_per_thread = subs_idxc(custom.elements_per_thread)
+        gpr_offsets = [
+            v.start for k, v in simplified_index.items() if v.start.has(GPR_NUM)
+        ]
+        assert len(gpr_offsets) == 1, "Expected only 1-Dim has gpr offsets"
+        gpr_offset_expr = gpr_offsets[0]
+        gpr_cur_base_offset = gpr_offset_expr.subs({GPR_NUM: 0})
+        cur_elem_id = 0
+        with custom.graph.inserting_before(operator):
+            ops_to_combine = []
+            for i in range(elements_per_thread):
```

Comment:
Don't we know the size of each chunk from the index sequence? If so, we can already calculate how many chunks are there by doing elements_per_thread / size. Then we can just iterate over the chunks and I think we can avoid the need to check the step difference altogether. What do you think?

Reply:
> Don't we know the size of each chunk from the index sequence

It is not really explicitly stated/queriable. We can do some computation to get the size of each chunk outside the loop. 

I think I'd do that actually, although it requires us to add some more code outside the main loop, but may be good since it will make the main/inner loop of this more robust.

Reply:
Sounds good thanks!

Reply:
Done, I think :)